cpp/src/arrow/array/builder_binary.h

```diff
@@ -67,7 +67,7 @@ class BaseBinaryBuilder
 
   Status Append(const uint8_t* value, offset_type length) {
     ARROW_RETURN_NOT_OK(Reserve(1));
-    ARROW_RETURN_NOT_OK(AppendNextOffset());
+    UnsafeAppendNextOffset();
```

Comment:
Hi, since `Append` is a public interface, how can user make sure that the offset is reserved before?